### PR TITLE
Add function to transform position

### PR DESCRIPTION
**Pull request by vessd**
## Description

Function to transform position from screen space into world space

## Additions

- `fn screen_to_world`

### CODE REVIEW

File: amethyst_utils/src/screen_to_world.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+//! Provides utilities for transform position.
+
+use amethyst_core::nalgebra::{Matrix4, Orthographic3, Vector3};
+use amethyst_renderer::ScreenDimensions;
+
+/// Transforms position from screen space into world space
+pub fn screen_to_world(
+    position: (f64, f64),
```
Comment (jojolepro) [inline amethyst_utils/src/screen_to_world.rs]: Can you rename this to screen_position?

File: amethyst_utils/src/screen_to_world.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+//! Provides utilities for transform position.
+
+use amethyst_core::nalgebra::{Matrix4, Orthographic3, Vector3};
+use amethyst_renderer::ScreenDimensions;
+
+/// Transforms position from screen space into world space
+pub fn screen_to_world(
+    screen_position: (f64, f64),
+    camera_translation: &Vector3<f32>,
+    camera_proj: &Matrix4<f32>,
```
Comment (Moxinilian) [inline amethyst_utils/src/screen_to_world.rs]: Wouldn't it make more sense here to directly take the Orthographic3 struct here?
It makes the user aware of what the parameter is, gives them the responsibility of providing a good projection and saves a cast when they already have the Orthographic3 struct.

Comment (vessd) [reply amethyst_utils/src/screen_to_world.rs]: Perhaps. I have little experience in development.
In my case it allows me not to write `use amethyst::core::::nalgebra::Orthographic3`.
And I probably would not understand where to get it, if I saw it the first time.
Right now, it’s enough for the user to pass `camera.proj` without any conversions.

Comment (Moxinilian) [reply amethyst_utils/src/screen_to_world.rs]: Well, I think the real issue here is that camera.proj is a Matrix4 and not an Orthographic3, as Orthographic3 anyway is just a wrapper for Matrix4. Let's leave it as is for now, it should be improved in the new renderer.

Comment (vessd) [reply amethyst_utils/src/screen_to_world.rs]: May be this should to take a link to the camera object for greater clarity

**Comment by Moxinilian** [comment]
The tests unexpectedly failed, but it probably is a glitch. This will not prevent merging once 24 hours will have passed.

File: amethyst_utils/src/screen_to_world.rs (hunk 2 above)
Comment (Moxinilian) [reply amethyst_utils/src/screen_to_world.rs]: Good idea. What you can do is in the doc just above explain that the matrix is the camera projection matrix field, or an Orthographic3 matrix.

Comment (vessd) [reply amethyst_utils/src/screen_to_world.rs]: Do you have an example of how I should format this?

**Comment by Moxinilian** [comment]
Actually, come to thing of it, why don't you make this a function on the Camera component?
So we could do something like `camera.coordinates_from_screen(screen_coords, screen_dim)`.
Maybe that would be more convenient.

**Comment by jojolepro** [comment]
+ the camera global transform as a parameter.

Making it a function of the camera makes sense.

The camera might be a child of another entity, so it would also make sense to use the global transform instead of the translation. However, this will make the math harder, so I don't know if we want that in this PR or not. We don't want to make you work too hard :D

**Comment by vessd** [comment]
Should I remove the translation argument then?

**Comment by Moxinilian** [comment]
Oh right, the translation. Hmmm. Maybe the function could take its own transform or something.
Ah that's a tricky one actually. The reason I would like it to be on the Camera component is because it is an obvious place to look for people coming from non-ECS places. We should look at how Unity is doing it with their ECS.

**Comment by Moxinilian** [comment]
Okay so apparently they're using globals, which is not very good. We should not do that.
I guess we should have it be a method on the Camera, to which we pass the camera's transform, the screen position and the ScreenDimension. What do you think?

File: amethyst_renderer/src/cam.rs [hunk 3]
```diff
@@ -76,6 +79,27 @@ impl Camera {
             std::f32::consts::FRAC_PI_3,
         ))
     }
+
+    /// Transforms position from screen space to camera space
```
Comment (jojolepro) [inline amethyst_renderer/src/cam.rs]:
```suggestion
    /// Transforms position from screen space to camera space
    /// This only works if the camera is not a child of another entity and if it is not rotated.
```

Comment (Moxinilian) [reply amethyst_renderer/src/cam.rs]: Why not use GlobalTransform then?

File: amethyst_renderer/src/cam.rs [hunk 4]
```diff
@@ -76,6 +79,22 @@ impl Camera {
...
+    pub fn position_from_screen(
+        &self,
+        screen_position: (f64, f64),
+        camera_transform: &GlobalTransform,
+        screen_dimensions: &ScreenDimensions,
+    ) -> (f32, f32) {
+        let screen_x = 2.0 * screen_position.0 as f32 / screen_dimensions.width() - 1.0;
+        let screen_y = 1.0 - 2.0 * screen_position.1 as f32 / screen_dimensions.height();
+        let screen_point = Point3::new(screen_x, screen_y, 0.0).to_homogeneous();
```
Comment (vessd) [inline amethyst_renderer/src/cam.rs]: I'm not sure about 0.0, but it works fine for me

Comment (Moxinilian) [reply amethyst_renderer/src/cam.rs]: What do you mean?

Comment (vessd) [reply amethyst_renderer/src/cam.rs]: I don't know if it's ok to set `0.0` for the z axis for `screen_point`

File: amethyst_renderer/src/cam.rs [hunk 5]
```diff
@@ -76,6 +79,22 @@ impl Camera {
...
+
+        let camera_point = camera_transform.0 * self.proj.try_inverse().unwrap() * screen_point;
```
Comment (Moxinilian) [inline amethyst_renderer/src/cam.rs]: Could you make this an `except` instead of an `unwrap`?

File: amethyst_renderer/src/cam.rs (hunk 4 above)
Comment (Moxinilian) [reply amethyst_renderer/src/cam.rs]: I believe it is, considering Z information only resides in the camera. Could you try it with a weird camera, like one with translated and rotated?

Comment (vessd) [reply amethyst_renderer/src/cam.rs]: Yes, I tested it with parameters
```
transform.set_xyz(17.0, -2.0, 1.0);
transform.set_rotation_euler(0.0, 0.0, 1.0);
```
It works fine

File: amethyst_renderer/src/cam.rs [hunk 6]
```diff
@@ -76,6 +79,24 @@ impl Camera {
...
+        screen_position: (f64, f64),
+        camera_transform: &GlobalTransform,
+        screen_dimensions: &ScreenDimensions,
+    ) -> (f32, f32) {
+        let screen_x = 2.0 * screen_position.0 as f32 / screen_dimensions.width() - 1.0;
+        let screen_y = 1.0 - 2.0 * screen_position.1 as f32 / screen_dimensions.height();
+        let screen_point = Point3::new(screen_x, screen_y, 0.0).to_homogeneous();
+
+        let camera_point = camera_transform.0
+            * self.proj.try_inverse().expect("Matrix is not invertible")
```
Comment (Moxinilian) [inline amethyst_renderer/src/cam.rs]:
```suggestion
            * self.proj.try_inverse().expect("Camera projection matrix is not invertible")
```

**Comment by Moxinilian** [comment]
Actually, while I'm at it, why not return a Vector3 instead of just X and Y? It can be useful when your camera is not on the Z = 0 plane. Also, why are the input coordinates f64? It could be a Vector2, even.

**Comment by vessd** [comment]
> Also, why are the input coordinates f64? It could be a Vector2, even.

Because the `InputHandler::mouse_position()` returns the `Option<(f64, f64)>`.

> Also, why are the input coordinates f64? It could be a Vector2, even.

Hmm, I don’t understand how it works clearly, do I need to take `Vec3` for this?
I mean, is it ok that `screen_point` is always equal to` 0.0` for the `z` plane for this? Or do i need to accept it as a parameter?
I didn't test the `z` coordinate, since I have no idea how to interpret it in my 2D game

**Comment by Moxinilian** [comment]
I think introducing a Z coordinate here as input would just shift how "deep" the click is made. It should not really matter here, we only are converting click position, not raycasting.

However, the Z output can be useful if your camera is in 3D space. As we are doing an inverted camera projection, we are effectively taking the screen plane at Z=0 and applying it to the 3D plane orthogonal to the camera's direction. That plane can be a Z=0 plane, but not necessarily, if the camera is in a weird position.

**Comment by vessd** [comment]
@Moxinilian ok, should I leave a f64 as an input argument?

**Comment by Moxinilian** [comment]
I think you should do as little casting as possible in the body of the function so the user does not have to do useless casting if they already work with f32. Set it to f32 in the parameter, and let the user do the casting.

**Comment by vessd** [comment]
I think it is ready to merge.
r? @Moxinilian @jojolepro @azriel91

File: amethyst_renderer/src/cam.rs [hunk 7]
```diff
@@ -76,6 +79,27 @@ impl Camera {
...
+        &self,
+        screen_position: Vector2<f32>,
+        camera_transform: &GlobalTransform,
+        screen_dimensions: &ScreenDimensions,
+    ) -> Vector3<f32> {
+        let screen_x = 2.0 * screen_position.x / screen_dimensions.width() - 1.0;
+        let screen_y = 1.0 - 2.0 * screen_position.y / screen_dimensions.height();
+        let screen_point = Point3::new(screen_x, screen_y, 0.0).to_homogeneous();
+
+        let matrix = camera_transform.0
```
Comment (Moxinilian) [inline amethyst_renderer/src/cam.rs]: Isn't this already a Vector3 or Point3?

File: amethyst_renderer/src/cam.rs [hunk 8]
```diff
@@ -76,6 +79,27 @@ impl Camera {
...
+            * self
+                .proj
+                .try_inverse()
+                .expect("Camera projection matrix is not invertible")
+            * screen_point;
+
+        Vector3::new(matrix.x, matrix.y, matrix.z)
```
Comment (Moxinilian) [inline amethyst_renderer/src/cam.rs]: I said Vector3 earlier, but I forgot nalgebra has a Point3 type which might make more sense here. Same applies for the input Vector2 with Point2.

File: amethyst_renderer/src/cam.rs (hunk 7 above)
Comment (vessd) [reply amethyst_renderer/src/cam.rs]: It's a `Matrix1x4`.
I haven't found an easy way to convert it to `Vector3` or `Point3`.

Comment (vessd) [reply amethyst_renderer/src/cam.rs]: Hm, I think it should be `Point3::from_homogeneous`

Comment (Moxinilian) [reply amethyst_renderer/src/cam.rs]: Ah you are correct, as it's a 1x4 we can't just cast it. I think your initial solution is good then.

**Comment by azriel91** [comment]
bors r=jojolepro,Moxinilian,azriel91

**Pull request merged**